src/main/java/de/azapps/kafkabackup/common/partition/cloud/S3PartitionWriter.java

```diff
@@ -1,76 +1,104 @@
 package de.azapps.kafkabackup.common.partition.cloud;
 
-import com.amazonaws.services.s3.model.ObjectMetadata;
 import de.azapps.kafkabackup.common.partition.PartitionException;
 import de.azapps.kafkabackup.common.partition.PartitionWriter;
 import de.azapps.kafkabackup.common.record.Record;
-import de.azapps.kafkabackup.common.record.RecordJSONSerde;
 import de.azapps.kafkabackup.storage.s3.AwsS3Service;
-import java.io.ByteArrayInputStream;
-import java.io.ByteArrayOutputStream;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.errors.RetriableException;
+
 import java.io.IOException;
-import lombok.Builder;
-import lombok.RequiredArgsConstructor;
+import java.util.LinkedList;
+import java.util.Queue;
 
-@RequiredArgsConstructor
-@Builder
+@Slf4j
 public class S3PartitionWriter implements PartitionWriter {
 
-  private final String bucketName;
-  private final int partition;
-
-  private final String topicName;
   private final AwsS3Service awsS3Service;
+  private final String bucketName;
+  private final TopicPartition topicPartition;
+  private final int maxBatchMessages;
+  private final long maxBatchTimeMs;
+  private final Queue<Record> buffer = new LinkedList<>();
 
-  private final RecordJSONSerde recordJSONSerde = new RecordJSONSerde();
+  private S3BatchWriter batchWriter;
+  private Long lastCommittableOffset = null; // Nothing written so far, so nothing to commit
 
-  private final long lastWrittenOffset = 0;
 
+  public S3PartitionWriter(AwsS3Service awsS3Service, String bucketName, TopicPartition tp, int maxBatchMessages, long maxBatchTimeMs) {
+    this.awsS3Service = awsS3Service;
+    this.bucketName = bucketName;
+    this.topicPartition = tp;
+    this.maxBatchMessages = maxBatchMessages;
+    this.maxBatchTimeMs = maxBatchTimeMs;
+  }
```

Comment:
if you don't have any special field value assignments, annotation @RequiredArgsConstructor, which was added before should be still valid 🤔